p2p/src/network/kad/request/p2p_network_kad_request_reducer.rs

```diff
@@ -1,52 +1,301 @@
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+
+                let stream_id = p2p_state
+                    .network
+                    .scheduler
+                    .connections
+                    .get(addr)
+                    .ok_or_else(|| format!("connection with {addr} not found"))
+                    .and_then(|conn| {
+                        conn.mux
+                            .as_ref()
+                            .map(|mux| (mux, conn.incoming))
+                            .ok_or_else(|| format!("multiplexing is not ready for {addr}"))
+                    })
+                    .and_then(|(P2pNetworkConnectionMuxState::Yamux(yamux), incoming)| {
+                        yamux
+                            .next_stream_id(crate::YamuxStreamKind::Kademlia, incoming)
+                            .ok_or_else(|| format!("cannot get next stream for {addr}"))
+                    })?;
+
+                dispatcher.push(P2pNetworkYamuxAction::OpenStream {
+                    addr: *addr,
+                    stream_id,
+                    stream_kind: crate::token::StreamKind::Discovery(
+                        crate::token::DiscoveryAlgorithm::Kademlia1_0_0,
+                    ),
+                });
+                dispatcher.push(P2pNetworkKadRequestAction::StreamIsCreating {
+                    peer_id: *peer_id,
+                    stream_id,
+                });
+                Ok(())
             }
-            P2pNetworkKadRequestAction::MuxReady { .. } => {}
             P2pNetworkKadRequestAction::StreamIsCreating { stream_id, .. } => {
-                self.status = super::P2pNetworkKadRequestStatus::WaitingForKadStream(*stream_id)
+                request_state.status = P2pNetworkKadRequestStatus::WaitingForKadStream(*stream_id);
+
+                Ok(())
             }
-            P2pNetworkKadRequestAction::StreamReady { .. } => {
-                let find_node =
-                    P2pNetworkKademliaRpcRequest::find_node(self.key).map_err(|e| e.to_string())?;
+            P2pNetworkKadRequestAction::StreamReady {
+                peer_id,
+                stream_id,
+                addr,
+            } => {
+                let find_node = match P2pNetworkKademliaRpcRequest::find_node(request_state.key) {
+                    Ok(find_node) => find_node,
+                    Err(error) => {
+                        bug_condition!(
+                            "P2pNetworkKadRequestAction::StreamReady invalid request key error: {error}"
+                        );
+                        return Ok(());
+                    }
+                };
+
                 let message = super::super::Message::from(&find_node);
-                self.status = quick_protobuf::serialize_into_vec(&message).map_or_else(
+                request_state.status = quick_protobuf::serialize_into_vec(&message).map_or_else(
```

Comment:
bug_condition?

Reply:
`request_state.key` is our PeerId so this should be bug_condition with us